http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/UriUtils.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal.utils;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.StringUtils;
+
+/**
+ * A dumbed-down version of Apache HTTP Components Client URIUtils.
+ */
+@SdkProtectedApi
+public class UriUtils {
+
+    private UriUtils() {
+    }
+
+    /**
+     * Converts an absolute URI to a relative URI by stripping out the scheme,
+     * authority, and fragment. In order to maintain compatibility with Apache
+     * HTTP Client behavior, opaque and non-absolute URIs are returned as-is.
+     * @param uri the uri
+     * @return the relativized URI
+     */
+    public static URI relativize(URI uri) {
+        Objects.requireNonNull(uri, "uri");
+        if (uri.isOpaque()) {
+            return uri;
+        }
+        try {
+            return new URI(null, null, null, -1, StringUtils.isEmpty(uri.getPath()) ? "/" : uri.getPath(), uri.getQuery(), null);

```

Comment:
`SdkRequest#getUri()` [does sort of the reverse of this](https://github.com/aws/aws-sdk-java-v2/blob/32bc73f98ca1e36fd4b6f8ce68733169c311eeb7/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java#L108-L121). Can we just copy the logic instead so we're doing less work?

Reply:
Copying the logic seems like it would be prone to breakage if someone changed how `SdkHttpRequest#getURI()` worked and forgot to update `UriUtils`. Better I think to let the request create a URI using consistent logic, and then operate on the result.

Reply:
I don't foresee it changing though, since it's pretty set in stone considering it's just following the RFC's. Any changes to the values returned by the functions like rawQueryParameters() would be a breaking change and not likely to happen either.

Reply:
@dagnir sorry but I don't think I can spend any more time working on this PR at this point. Feel free to make the change when bringing it in, if you're happy enough with the other changes to accept the PR.

Reply:
No worries! Thank you very much for the work you put into this PR, we really appreciate it. I will go ahead and update the PR and then merge it.